security/pkg/pki/ra/common.go

```diff
@@ -89,6 +99,69 @@ func ValidateCSR(csrPEM []byte, subjectIDs []string) bool {
 	return true
 }
 
+// basicConstraints is a structure that represents the ASN.1 encoding of the
+// BasicConstraints extension.
+// The structure is borrowed from
+// https://github.com/golang/go/blob/master/src/crypto/x509/x509.go#L975
+type basicConstraints struct {
+	IsCA       bool `asn1:"optional"`
+	MaxPathLen int  `asn1:"optional,default:-1"`
+}
+
+func checkIsCAExtension(exts []pkix.Extension) (bool, error) {
```

Comment:
I think this is too permissive and also risky.

We create the CSR in ztunnel or istio-agent (or istiod for dns certs - but those don't need validation since istiod creates them).

There is no point in accepting certs that have extensions unless we decide to create such CSRs. We don't have a requirement to support arbitrary CSRs. 

I don't think ztunnel or agent add extensions - so just drop anything that has an extension.

Reply:
I worry this is a bit to restrictive: CRLs for example are specified as an extension IIUC; why wouldn't we allow that usage?

Reply:
Who would generate CRLs, and how would we validate them ? 

I don't mind supporting CRLs or any other extension: it needs some design doc to explain how they'll be used, changes in Agent and ZTunnel to add them to the CSR and then changes to validator to allow them. 

Approving CSRs created by some other entities ( not agent/ztunnel) with random features is not supported with Istio CA - which just create a new cert template matching what agent needs and nothing else. 

The problem is that X509 has a lot of options that may have different impact - and for the purpose of Istio we don't need the RA path to have a very different behavior than the normal CA path. 

Certificate validation is supposed to be very restrictive - I don't think many CAs ( ACME or others ) allow random/unexpected stuff in the cert. If we have a feature to support CRL - it should be clearly defined and we should also add it in the normal CA path.

Reply:
> and for the purpose of Istio we don't need the RA path to have a very different behavior than the normal CA path.

Correct me if I'm wrong (I don't spend most of my time with certs), but isn't the whole purpose of having a separate CA vs. RA so that you can have the entity that authenticates the request be different from the entity that actually mints the certificate? If an organization has a requirement that all of their certs have a certain extension set, why should istiod (acting as an RA) care?

Reply:
IIUC the basic problem is that `istiod` can't actually act as a particularly robust RA, so we have to fence of potential abuses of the CSR model in other ways (like rejecting CSRs that look funny).